aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+
+#include <ATen/div_rtn.h>
+#include <algorithm>
+#include <tuple>
+#include "ATen/ATen.h"
+#include "ATen/native/im2col.h"
+#include "ATen/native/vol2col.h"
+#include "TH/THBlasUtils.h"
+
+#define TORCH_CHECK_DIM_SIZE(T, DIM, DIM_SIZE, SIZE) \
+  TORCH_CHECK(                                       \
+      T.dim() == DIM && T.size(DIM_SIZE) == SIZE,    \
+      "Need " #T " of dimension ",                   \
+      DIM,                                           \
+      " and " #T ".size[",                           \
+      DIM_SIZE,                                      \
+      "] == ",                                       \
+      SIZE,                                          \
+      " but got input to be of shape ",              \
+      T.sizes())
+
+/*
+  Some convenience macros
+ */
+
+#define INSERT_BATCH_DIMENSION(A, SIZE)        \
+  if (A.defined()) {                           \
+    auto new_sizes = A.sizes().vec();          \
+    new_sizes.insert(new_sizes.begin(), SIZE); \
+    A.resize_(new_sizes);                      \
+  }
+
+#define DROP_BATCH_DIMENSION(A)    \
+  if (A.defined()) {               \
+    A.resize_(A.sizes().slice(1)); \
+  }
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline bool all_positive(IntArrayRef& arr) {
+  return std::all_of(
+      arr.begin(), arr.end(), [](int64_t item) { return item > 0; });
+}
+
+#define OUTPUTSIZE(INDEX, DIM)                                       \
+  (input.size((INDEX) + input.dim() - (DIM)) + 2 * pad_size[INDEX] - \
+   (dilation_size[INDEX] * (kernel_size[INDEX] - 1) + 1)) /          \
+          stride_size[INDEX] +                                       \
+      1
+
+void conv_dilated_shape_check(
```

Comment:
Meta-point about all this shape checking nonsense: `ATen/native/Convolution.cpp` has *yet another copy* of this shape checking (in fact, I think we have like N copies of convolution shape checking, one per implementation we have, and we have a lot of implementations). So if you want to stop spending time fixing this port, and instead just get them to use the same shape checking code, that would also be a reasonable way to solve the problems in this block of code.

Reply:
I looked into Convolution.cpp argument checking and concluded that due to different features in Convolution and DilatedConvolution it becomes cumbersome to reuse this code.
So, I went ahead and fixed shape checking for DilatedConvolution which is now also shared between cpu and cuda versions of the conv_dilated functions.